### ISSUE TITLE
User redirected to account page when trying to checkout

### ISSUE THREAD

**Issue by mattcoz** [magento/magento2#12119]
### Preconditions
1. Magento 2.2
2. PHP 7.0.10

### Steps to reproduce
1. Log in
2. Add item to cart
3. Go to checkout

### Expected result
1. Able to proceed through checkout

### Actual result
1. User is redirected to account page

The cause seems to be a "401 Unauthorized" response to /rest/default/V1/carts/mine/estimate-shipping-methods-by-address-id

**Comment by mattcoz** [comment]
I wrote a plugin to debug the parameters and result of the isAllowed function of PolicyInterface, here is what I got:

```
[2017-11-08 21:15:18] main.DEBUG: beforeIsAllowed: null, "self", null. [] []
[2017-11-08 21:15:18] main.DEBUG: afterIsAllowed: false. [] []
```

Does this help?

**Comment by mattcoz** [comment]
I've tried disabling all custom modules and third party modules, still get the same behavior.

**Comment by magento-engcom-team** [comment]
@mattcoz, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. If you'd like to update it, please reopen the issue.
We tested the issue on 2.3.0-dev, 2.1.10, 2.2.1

**Comment by cotardo18** [comment]
hi @mattcoz , could you fix the issue?

**Comment by mattcoz** [comment]
I'm getting back to this now, still get the same behavior after updating to 2.2.2

**Comment by mattcoz** [comment]
I extended my debugging and found that the current user is not being picked up in the user context, the user id is 0 and the user type is GUEST.

**Comment by mattcoz** [comment]
I think I fixed it, the session was getting screwed up in one line of my code that seemed to be completely unrelated. So you can leave this closed.

**Comment by sambolek** [comment]
Hi @mattcoz - could you please provide an example of how your code was screwing up sessions? We're experiencing the same issue (from what I can see, there are others as well), so interested to see if this can be related to some part of 3rd party/custom code, as we have no errors logged anywhere, but customers get logged out constantly.

**Comment by jrocchiasasdesmelie** [comment]
Hi @mattcoz  - same problem here. could you help us please ? :smile:

**Comment by mattcoz** [comment]
I was calling session_start() before the bootstrap code in order to check a session variable. I think because the session was already initialized that Magento was bypassing some of its initialization code. I added a call to session_abort() after I was done with it and the problem went away.

https://github.com/magento/magento2/blob/4d4e88a89c5fa525c5ac442a1169ec71d7f8dda0/lib/internal/Magento/Framework/Session/SessionManager.php#L169

**Comment by sambolek** [comment]
Thanks Matt, that was very helpful.

It turned out to be an issue with one of the Mirasvit extensions (Rewards) that was checking/modifying sessions too early (on controller_action_predispatch event).

**Comment by lennartbee** [comment]
@sambolek we are having the same issue and we are using the Mirasvit Rewards module. The only controller_action_predispatch event I can find in the Mirasvit Module is in checkout_onepage_success. Could you please specify which file you edited?

**Comment by lennartbee** [comment]
I found it already, in our case it wasn't caused by the Mirasvit module. https://github.com/magento/magento2/commit/f82a17507acfa821e2bea543a99f072e1de1252f fixed it for us.

**Comment by antbates91** [comment]
@magento-engcom-team I can replicate this on 2.2.2, this issue should be reopened.

It does appear to only happen on a few occasions, probably 1 in 10 times but it is an issue.

### Preconditions
1. Magento 2.2
2. PHP 7.0.10

### Steps to reproduce
1. Log in
2. Add item to cart
3. Go to checkout
4. If the checkout loads, empty the basket, sign out and repeat steps 1 -> 3 again.

### Expected result
1. Able to proceed through checkout

### Actual result
1. User is redirected to account page

**Comment by akios111** [comment]
> I was calling session_start() before the bootstrap code in order to check a session variable. I think because the session was already initialized that Magento was bypassing some of its initialization code. I added a call to session_abort() after I was done with it and the problem went away.
> 
> [magento2/lib/internal/Magento/Framework/Session/SessionManager.php](https://github.com/magento/magento2/blob/4d4e88a89c5fa525c5ac442a1169ec71d7f8dda0/lib/internal/Magento/Framework/Session/SessionManager.php#L169)
> 
> Line 169 in [4d4e88a](/magento/magento2/commit/4d4e88a89c5fa525c5ac442a1169ec71d7f8dda0)
> 
>          if (!$this->isSessionExists()) {

Hello,could you please explain what you code you put and where to fix this issue?
It's happening to me also

**Comment by goivvy** [comment]
He had an issue like this and oh boy. It was hard to figure out until I noticed a custom theme file had been edited. Long story short, it's most like your custom theme or your custom extension. Turn off all custom modules and switch to Luma. Then if it starts working, turn on modules one by one to see which one breaks the site. If they all work fine then it's your custom theme that needs closer inspection. Cheers!

**Comment by nimbus2300** [comment]
Hi,

I've just spend the best part of a day debugging an issue with this end-point and I thought I'd share my findings in case it saves anyone else some time.

TL;DR: Caused by a third party social-login plugin that 'logged in' the wrong user (same email address, but wrong store view).

**The symptoms:**
* Sign in to user account that has an delivery address registered with it
* Add something to cart
* Go to checkout, and upon loading the checkout page, it redirects to the Account page

**What is happening in the background:**
* If the software recognises it is a known/logged in user, it calls /rest/default/V1/carts/mine/estimate-shipping-methods-by-address-id on the checkout page, to grab a list of shipping estimates for their known addresses.
* This end point authenticates the account holding user that is logged in to your website. If it can't find the user, or can't authenticate it returns 401 Unauthorised.

**Why this was happening in my store**
* My store is a multi-store setup.
* Note that each user account has an "Account Created In" field describing the store view that account belongs to.
* In my case, I was using a third party social plugin, Mageplaza, which had a google login button
* The login button was matching my email address to a user account in site A, even though I was using site B. What it should have done, is create a new user for site B.
* I was browsing the website Site B with a 'logged in' user from site A.
* When the Ajax call was made, the authentication rightly failed.
* The error mechanism sends to the account page (presumably to re-authenicate).

**Summary**
In my case this is caused by a faulty Google Login account matching mechanism in the version of the Mageplaza Social Login button I was using. After authenticating with Google, it grabbed the incorrect user account (from another store) and set this as the active user. I'm going to upgrade/fix this.
